src/lib/shell/Engine.h

```diff
@@ -93,20 +100,78 @@ typedef const struct shell_command shell_command_t;
  */
 typedef CHIP_ERROR shell_command_iterator_t(shell_command_t * command, void * arg);
 
+class Engine;
+
+/**
+ * A map of the Engine instances and the command prefix that they correspond to.
+ * The struct itself is a linked list node.
+ *
+ *  prefix:     The command prefix (until n-1th token of full command) e.g. "device",
+ *              or "dns resolve".
+ *  enginev:    An array of the shells that have registered commands under the preifx.
+ *  enginec:    The number of shells that have registered commands under the preifx.
+ *  next:       The next shell map in the list.
+ *
+ */
+typedef struct shell_map
+{
+    const char * prefix;
+    Engine * enginev[CHIP_SHELL_MAX_MODULES];
+    size_t enginec = 0;
+    shell_map * next;
+} shell_map_t;
+
+/**
+ * A context object for passing request and receiving results with GetCmdCompletion.
+ *
+ *  line_buf:           The user input command to request for completion. If the applications prepends
+ *                      "matter " to all matter commands, please remove the "matter " prefix and only use
+ *                      the string after it.
+ *  ret_prefix:         The returned command prefix (up until the last space, not included).
+ *  cmdv:               The returned command completion candidates under the prefix in "ret_prefix".
+ *  cmdc:               The returned number of command completion candidates in cmdv.
+ *
+ * Initialization:
+ *
+ *  cmd_completion_context context = cmd_completion_context("dns browse c");
+ *
+ */
+typedef struct cmd_completion_context
+{
+    const char * line_buf;
+    const char * ret_prefix;
+    shell_command_t * cmdv[CHIP_SHELL_MAX_CMD_COMPLETIONS];
+    size_t cmdc = 0;
+
+    cmd_completion_context(){};
+    cmd_completion_context(const char * _line_buf) { line_buf = _line_buf; };
+} cmd_completion_context;
+
 class Engine
 {
 protected:
     static Engine theEngineRoot;
+    static shell_map_t * theShellMapListHead;
+    shell_command_t * _commands[CHIP_SHELL_MAX_MODULES];
+    unsigned _commandCount;
 
-    shell_command_t * _commandSet[CHIP_SHELL_MAX_MODULES];
-    unsigned _commandSetSize[CHIP_SHELL_MAX_MODULES];
-    unsigned _commandSetCount;
+    static void InsertShellMap(char const * prefix, Engine * shell);
 
 public:
-    Engine() {}
+    Engine(){};
```

Comment:
supernit: remove the `;` change here and below. There should be no ; after method implementations